rest/route/project.go

```diff
@@ -883,83 +884,205 @@ func (h *getProjectVersionsHandler) Parse(ctx context.Context, r *http.Request)
-			return errors.Wrap(err, "unmarshalling JSON request body into version options")
+	opts, err := parseGetVersionsOptions(b, params)
+	if err != nil {
+		return err
+	}
+	h.opts = opts.GetVersionsOptions
+	return nil
+}
+
+func (h *getProjectVersionsHandler) Run(ctx context.Context) gimlet.Responder {
+	versions, err := data.GetProjectVersionsWithOptions(h.projectName, h.opts)
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err, "getting versions for project '%s'", h.projectName))
+	}
+
+	resp, err := gimlet.NewBasicResponder(http.StatusOK, gimlet.JSON, versions)
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "constructing response"))
+	}
+
+	if len(versions) >= h.opts.Limit {
+		err = resp.SetPages(&gimlet.ResponsePages{
+			Next: &gimlet.Page{
+				Relation:        "next",
+				LimitQueryParam: "limit",
+				KeyQueryParam:   "start",
+				BaseURL:         h.url,
+				Key:             strconv.Itoa(versions[len(versions)-1].Order),
+				Limit:           h.opts.Limit,
+			},
+		})
+
+		if err != nil {
+			return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "paginating response"))
 		}
 	}
 
-	if h.opts.IncludeTasks && !h.opts.IncludeBuilds {
-		return errors.New("cannot include tasks without builds")
+	return resp
+}
+
+// POST /rest/v2/projects/{project_id}/versions
+
+// modifyProjectVersionsHandler is a RequestHandler for setting the priority of versions.
+type modifyProjectVersionsHandler struct {
+	projectId string
+	url       string
+	opts      dbModel.ModifyVersionsOptions
+}
+
+func makeModifyProjectVersionsHandler(url string) gimlet.RouteHandler {
+	return &modifyProjectVersionsHandler{url: url}
+}
+
+func (h *modifyProjectVersionsHandler) Factory() gimlet.RouteHandler {
+	return &modifyProjectVersionsHandler{url: h.url}
+}
+
+func (h *modifyProjectVersionsHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.projectId = gimlet.GetVars(r)["project_id"]
+	params := r.URL.Query()
+
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return errors.Wrap(err, "reading request body")
+	}
+	opts, err := parseGetVersionsOptions(b, params)
+	if err != nil {
+		return err
+	}
+	if opts.Priority == nil {
+		return errors.New("must specify a priority")
+	}
+	h.opts = *opts
+	return nil
+}
+
+func (h *modifyProjectVersionsHandler) Run(ctx context.Context) gimlet.Responder {
+	user := MustHaveUser(ctx)
+	priority := utility.FromInt64Ptr(h.opts.Priority)
+	versions, err := data.GetProjectVersionsWithOptions(h.projectId, h.opts.GetVersionsOptions)
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err, "getting versions for project '%s'", h.projectId))
+	}
+	var versionIds []string
+	for _, v := range versions {
+		versionIds = append(versionIds, utility.FromStringPtr(v.Id))
+	}
+
+	// Check for a valid priority and perform the update.
+	if ok := validPriority(priority, h.projectId, user); !ok {
```

Comment:
Can we check this before doing the version query?

Reply:
Sure